src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
+    {
+        private int _createdCount;
+        private int _disposedCount;
+
+        static DisposeScopeManager()
+        {
+            torch.Tensor.OnTensorCreateCallback.Add(TryRegisterInDisposeScope);
+            torch.Tensor.OnTensorDisposeCallback.Add(TryUnregisterInDisposeScope);
+        }
+
+        [ThreadStatic] private static DisposeScopeManager _singleton;
+
+        public static DisposeScopeManager Singleton = (_singleton ??= new DisposeScopeManager());
+
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        internal IDisposable CurrentlyDisposing { get; set; }
+
+        public int CreatedCount => _createdCount;
+        public int DisposedCount => _disposedCount;
+        public int TotalCount => _createdCount - _disposedCount;
+
+        public static long TotalAllocatedSize => Singleton.InnerTotalAllocatedSize;
+
+        internal static void TryUnregisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryUnregisterInDisposeScope(disposable);
+
+        internal static void TryRegisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryRegisterInDisposeScope(disposable);
+
+        private long InnerTotalAllocatedSize {
+            get {
+                long size = 0;
+                lock (this) {
+                    foreach (var disposeScope in DisposeScopeStack) {
+                        size += disposeScope.TotalAllocatedSize;
+                    }
+                }
+
+                return size;
+            }
+        }
+
+        private void InnerTryRegisterInDisposeScope(IDisposable disposable)
+        {
+            Interlocked.Increment(ref _createdCount);
+
+            lock (this) {
+                if (DisposeScopeStack.Count > 0) {
+                    DisposeScopeStack.Peek().Include(disposable);
+                }
+            }
+        }
+
+        private void InnerTryUnregisterInDisposeScope(IDisposable disposable)
+        {
+            // It's us! We're disposing of it.
+            if (disposable == CurrentlyDisposing) {
```

Comment:
Can you describe a scenario when this could happen?

Reply:
It was to prevent a loop where us disposing a tensor made the tensor contact us to let us know it was being disposed. Since we already knew we were disposing it, we could save a roundtrip. I've solved it a different way now!